### PR TITLE
Bump k8s packages from 0.23.6 to 0.24.0

### PR DESCRIPTION
**Pull request by siggy**
This required modifying the `kube.Factory` interface to conform to
changes in k8s' `cmdutil.Factory` interface:
https://github.com/kubernetes/kubernetes/commit/fe3772890f650f9bcf020b43dc5a51fab0fa17f4

Signed-off-by: Andrew Seigner <andrew@sig.gy>

**What this PR does / why we need it**:

Bumps the k8s dependencies from 0.23.6 to 0.24.0.

**Special notes for your reviewer**:

This dependency bump required modifying source code beyond `go.mod`/`go.sum`.

**If applicable**:
- [ ] this PR contains documentation
- [ ] this PR contains unit tests
- [ ] this PR has been tested for backwards compatibility

### CODE REVIEW

File: pkg/kube/factory.go [hunk 1]
```diff
@@ -28,11 +30,19 @@ import (
 type Factory interface {
 	// ToRawKubeConfigLoader return kubeconfig loader as-is
 	ToRawKubeConfigLoader() clientcmd.ClientConfig
+
+	// DynamicClient returns a dynamic client ready for use
+	DynamicClient() (dynamic.Interface, error)
```
Comment (mattfarina) [inline pkg/kube/factory.go:35]: The [Helm HIP that covers backwards compatibility](https://github.com/helm/community/blob/main/hips/hip-0004.md) refers to the[ Go project stance and direction](https://go.dev/blog/module-compatibility#working-with-interfaces) on this which is that modifying interfaces is backward compatibility breaking. So, we add new interfaces for these cases. Could you please put them on a new interface and use the method in the linked Go docs to work with the new interface.

**Review by mattfarina — CHANGES_REQUESTED** [review]

File: pkg/kube/factory.go [hunk 2]
```diff
@@ -28,11 +30,19 @@ import (
...
 	// KubernetesClientSet gives you back an external clientset
 	KubernetesClientSet() (*kubernetes.Clientset, error)
+
 	// NewBuilder returns an object that assists in loading objects from both disk and the server
 	// and which implements the common patterns for CLI interactions with generic resources.
 	NewBuilder() *resource.Builder
+
 	// Returns a schema that can validate objects stored on disk.
-	Validator(validate bool) (validation.Schema, error)
+	Validator(validationDirective string, verifier *resource.QueryParamVerifier) (validation.Schema, error)
```
Comment (mattfarina) [inline pkg/kube/factory.go:45]: I missed it in my first review. This interface can't change. It would be backwards compatibility breaking.

**Review by mattfarina — CHANGES_REQUESTED** [review]

**Comment by mattfarina** [comment]
ok, it seems `Factory` is a mirror (or sorts) for an upstream `Factory` in kubectl. I did some more digging. I'm checking with other maintainers on how we want to handle this. Something needs to change in a way that's technically breaking.

File: pkg/kube/factory.go (hunk 1 above)
Comment (siggy) [reply pkg/kube/factory.go:35]: @mattfarina Thanks for the review and links. I'm a bit unsure on the best way forward. I assume Helm's `kube.Factory` interface was built to mimic/wrap k8s' `cmdutil.Factory`. In `v0.24.0`, `cmdutil.Factory`'s `Validator` method signature was modified. For `kube.Factory` to maintain backwards compatibility, we must continue supporting the older `Validator` signature, which no longer exists in k8s/k8s. To do this, I've introduced a `NewFactoryShim()` helper, that takes a `cmdutil.Factory` and returns a `kube.Factory`.

So a consumer that was previously doing this:
```go
&kube.Client{
	Factory: cmdutil.NewFactory(getter),
}
```
... now does this:
```go
&Client{
	Factory: NewFactoryShim(cmdutil.NewFactory(getter)),
}
```

Here is the full implementation:
https://github.com/siggy/helm/compare/main...siggy:siggy/k8s-v0-24-0-shim

An alternate (arguably less backwards-compatible) approach, would be to simply remove `kube.Factory` in favor of `cmdutil.Factory`. This approach would require no code changes for the consumer in the example above:
https://github.com/siggy/helm/compare/main...siggy:siggy/k8s-v0-24-0-remove

This all perhaps boils down to whether we break consumers that are implementing a `kube.Factory` directly, vs. breaking those that call `cmdutil.NewFactory(getter)` to get a `cmdutil.Factory`. In my own project, we're doing the latter.

Comment (siggy) [reply pkg/kube/factory.go:35]: ....aaand I just saw your subsequent review comments after posting this. Sounds like we're converging on an approach. Standing by.

File: pkg/kube/client.go [hunk 3]
```diff
@@ -192,7 +192,18 @@ func (c *Client) newBuilder() *resource.Builder {
...
+		validationDirective = metav1.FieldValidationStrict
+	}
+
+	dynamicClient, err := c.Factory.DynamicClient()
+	if err != nil {
+		return nil, err
+	}
+
+	verifier := resource.NewQueryParamVerifier(dynamicClient, c.Factory.OpenAPIGetter(), resource.QueryParamFieldValidation)
+	schema, err := c.Factory.Validator(validationDirective, verifier)
```
Comment (mattfarina) [inline pkg/kube/client.go:206]: Note to other Helm reviewers, this is roughly how kubectl does it.

**Review by mattfarina — APPROVED** [review]
To document why I'm approving this breaking change.

The Factory interface in Helm is a mirror of the Factory interface in kubectl. Kubernetes makes breaking changes in minor versions. This is ultimately a situation of Helm leaking the Kubernetes breaking change. We have a policy that we can't fix for Kubernetes breaking changes. So, this interface needs to change.

Why not use the kubectl interface directly? As I understand it, we do this to minimize the breaking surface area. For example, we don't need every function on the interface so addition of new ones don't need to cause breaking changes. This is a situation where we need to bubble up the change.

As a follow-up, I'll document this in more detail.

**Review by hickeyma — APPROVED** [review]
As the `Validator` signature changed in Kubernetes 1.24, there is no alternative but to break the `Factory` interface in Helm. Thanks @siggy for pushing the changes.

**Comment by hickeyma** [comment]
> Why not use the kubectl interface directly? As I understand it, we do this to minimize the breaking surface area. For example, we don't need every function on the interface so addition of new ones don't need to cause breaking changes. This is a situation where we need to bubble up the change.
As a follow-up, I'll document this in more detail.

@mattfarina I think it would be great to document this before the reasoning behind this is lost.

**Pull request merged**